### DISCUSSION TITLE
SQL Book

### DISCUSSION THREAD

**Discussion by textbrowser** [textbrowser/biblioteq#179] [Show and tell]
https://nostarch.com/practical-sql-2nd-edition

**Comment by meteos77** [comment]
In order to create an annual report for the year I am starting to put the SQL into practice ( Afin de créer un rapport annuel pour l'année je commence a mettre en pratique le SQL)

- Number of documents in the biblioteQ (on the day of the request) ( Nombre de documents dans la biblioteQ (au jour de la requete) )
**SELECT COUNT (myoid) FROM book**

- Number of loans in 2021 ( Nombre de prêts en 2021 )
**SELECT COUNT (copyid) FROM item_borrower WHERE reserved_date like ('%/%/2021')**

- Number of members in the database (total)  ( Nombre de membres dans la base (total) )
**SELECT COUNT (memberid) FROM member**

- Number of active members in 2021 (at least one reservation is made) ( Nombre de membres actif en 2021 (on fait au moins une réservation) )
**SELECT COUNT (DISTINCT memberid) FROM member_history WHERE reserved_date like ('%/%/2021')**

**Reply by meteos77** [reply]
Allows you to export the book database in the same field order as the CSV import proposed in type "Model 1 (BiblioteQ books (all fields) - 33 fields".

Permet d'exporter la base des livres dans le même ordre de champs que l'import CSV proposer en type "Modèle 1 (Livres BiblioteQ (tous les champs) - 33 champs"

**SELECT accession_number, alternate_id_1, author, binding_type, callnumber, category, condition, date_of_reform, description, deweynumber, edition, id, isbn13, keyword, language, lccontrolnumber, location, marc_tags, monetary_units, multivolume_set_isbn, origin, originality, pdate, place, price, publisher,  purchase_date, quantity, series_title, target_audience, title, url, volume_number FROM book**

**Comment by meteos77** [comment]
from here the queries work if the location field is filled in ( à partir d'ici les requêtes fonctionnent si le champ categorie est renseigné )

Have statistics on the type of documents you have in the library ( Avoir des statistiques sur le type de documents que l'on a dans la bibliothèque )

- Number of documents without category ( Nombre de documents dans aucune catégorie )
**SELECT  count (myoid) from book WHERE location like ('inconnue')**

- Number of documents in the category "Adult Comics" ( Nombre de documents dans la catégorie "BDA-Bande Déssinée Adultes" )
**SELECT  count (myoid) from book WHERE location like ('BDA-Bande Déssinée Adultes')**

- number of documents in the category Docmumentaries "Class 000 - Documentaries Computer, information, general works" ( nombre de documents dans la catégorie Docmumentaires "Classe 000 - Documentaires Informatique, information, ouvrages généraux" )
**SELECT  count (myoid) from book WHERE location like ('Classe 000%')**

- Number of documents in the category "R - Novel ( Nombre de documents dans la catégorie "R - Roman" )
**SELECT  count (myoid) from book WHERE location like ('R - Roman')**

- Number of documents in the category "Detective novel" ( Nombre de documents dans la catégorie "RP - Roman Policier" )
**SELECT  count (myoid) from book WHERE location like ('RP - Roman Policier')**

**Reply by meteos77** [reply]
- To list documents that do not have a cover (Pour lister les documents qui n'ont pas de couverture)
**SELECT book.callnumber, book.title, book.author, book.isbn13, book.myoid, book.type, book.front_cover FROM book WHERE front_cover IS NULL**

**Comment by meteos77** [comment]
some ideas but still too complicated for my SQL knowledge :-)
- Average number of loans per reader ( Moyenne du nombre d'emprunts par lecteurs )
- Average borrowing time ( Temps moyen d'emprunt )
- Type of documents most borrowed ( Type de documents les plus empruntés )

**Reply by textbrowser** [reply]
SELECT (SELECT COUNT(*) FROM item_borrower) / (SELECT COUNT(*) FROM member) (division by zero possible)
SELECT COUNT(*), type FROM item_borrower GROUP BY type ORDER BY 1 DESC

Borrowing time requires date conversions.

**Comment by meteos77** [comment]
to search for duplicates in the title field (pour rechercher les doublons sur le champ titre affiche leur nombre)

**SELECT COUNT(*) AS nbr_doublon, accession_number, title FROM book GROUP BY title HAVING COUNT(\*) > 1**

**SELECT COUNT(*) AS nbr_doublon, * FROM book GROUP BY title HAVING COUNT(\*) > 1**

Fnding all entries which have duplicate titles (Recherche de toutes les entrées qui ont des titres en double affiche les enregistrements)
  
  **SELECT accession_number, myoid FROM book WHERE title IN (SELECT title FROM book GROUP BY title HAVING COUNT(title) > 1);**

Finding all entries with duplicate accession numbers (Recherche de toutes les entrées avec des numéros d'accession en double)
**SELECT accession_number, myoid FROM book WHERE accession_number IN (SELECT accession_number FROM book GROUP BY accession_number HAVING COUNT(accession_number) > 1);**



to search for duplicates on the accession_number field and sort and display them all (pour rechercher les doublons sur le champ accession_number et les trier et les afficher les enregistrements)

**SELECT a.\* FROM book AS a JOIN (SELECT accession_number, COUNT() FROM book GROUP BY accession_number HAVING COUNT(*) > 1 ) AS b ON a.accession_number = b.accession_number ORDER BY accession_number**

**Comment by meteos77** [comment]
To find out if an exceptional value exists in the target audience (Pour rechercher si une valeur exceptionnelle existe dans public cible)
**SELECT DISTINCT target_audience FROM book**

to create the information of the labels of the books via export in csv and exploitation in glabels (pour créer les infos des étiquettes de la tranche des livres via export en csv et exploitation dans glabels)

**SELECT SUBSTR(location,1 , INSTR(location,'-')-1 ) AS Typ, UPPER(SUBSTR(author,1,3)) AS Aut, accession_number AS Num FROM book;**

[exemple-etiquettes.pdf](https://github.com/textbrowser/biblioteq/files/10156542/exemple-etiquettes.pdf)

**Comment by meteos77** [comment]
requête de base pour les permanences (Touche F6) - Permet d'avoir les informations indispensables dans la vue tableau
**SELECT accession_number AS "Référence", title AS "Titre",author AS "Auteur", isbn13 AS "ISBN", location AS "Localisation", callnumber AS "Etiquette", target_audience AS "Public",originality AS "Origine",myoid,type from book ORDER BY cast(accession_number as real) ASC**

**Comment by meteos77** [comment]
45 days after the document is returned, the history is removed. //45 jours après le retour du document, l'historique est enlevé.
`DELETE FROM member_history WHERE myoid IN (SELECT myoid FROM member_history  WHERE SUBSTR(returned_date, 7, 4) || '-' || SUBSTR(returned_date, 1, 2) || '-' || SUBSTR(returned_date, 4, 2) BETWEEN (SELECT DATE ('2000-01-01'))  AND (SELECT DATE ('now' ,'-45 day')));`

**Comment by meteos77** [comment]
1 year after the last contribution the inactive member is deleted (calculation: current year -1 years, which is less than the result).
ex : 2023 - 1 = 20222 so 2021 is deleted.

1 an après la dernière cotisation le membre inactif est supprimé (calcul : année actuelle -1 ans , ce qui est inférieur au résultat).
ex : 2023 - 1 = 20222 donc 2021 est supprimé.

`DELETE FROM member WHERE memberid IN (SELECT memberid FROM member WHERE SUBSTR(expiration_date, 7, 4)  < (SELECT SUBSTR(DATE('now', '-1 year'), 1, 4)));`

Pour obtenir la liste avant suppression
`SELECT last_name AS "NOM", first_name AS "Prénom" ,SUBSTR(expiration_date,7,8) || '-' || SUBSTR(expiration_date,1,2) || '-' || SUBSTR(expiration_date,4,2) AS "Date d'expiration", memberid AS "Identifiant" FROM member WHERE memberid IN (SELECT memberid FROM member WHERE SUBSTR(expiration_date, 7, 4) < (SELECT SUBSTR(DATE('now', '-1 year'), 1, 4)));`

**Reply by textbrowser** [reply]
Remember to first mark all reserved items as being available before deleting a member. The member table has this trigger:

CREATE TRIGGER member_history_trigger AFTER DELETE ON member
FOR EACH row
BEGIN
    DELETE FROM member_history WHERE memberid = old.memberid;
END;

The trigger does not remove item_borrower entries. Should it?

**Reply by textbrowser** [reply]
![bq-1](https://user-images.githubusercontent.com/10701156/233876557-e174749f-3d92-422d-9b13-ac45449859db.png)

**Comment by textbrowser** [comment]
New trigger.

https://github.com/textbrowser/biblioteq/commit/5a4229843dfd82bae52e45e52cb85ac31f00bc86

**Comment by meteos77** [comment]
Find out which books have been borrowed by whom? /  Trouver les livres et empruntés par qui ?
specify access number after the 2 INs / préciser le numero d'acces après les 2 IN

`SELECT book.callnumber, book.title, book.author, book.isbn13, book.myoid, book.type, member.last_name, item_borrower.duedate FROM book LEFT JOIN item_borrower ON book.myoid = item_borrower.item_oid, member WHERE book.accession_number IN ('7098','7095','7090') AND item_borrower.memberid = member.memberid UNION SELECT book.callnumber, book.title, book.author, book.isbn13, book.myoid, book.type, 'Disponible' AS last_name, item_borrower.duedate FROM book LEFT JOIN item_borrower ON book.myoid = item_borrower.item_oid WHERE book.accession_number IN ('7098','7095','7090') and item_borrower.item_oid IS NULL`

**Comment by meteos77** [comment]
to update the callnumber from location and author / pour mettre à jour la cote (callnumber) depuis emplacement (location) et auteur (author)
`UPDATE book SET callnumber = (SUBSTR(location,0, INSTR(location, '-'))  || ' ' || SUBSTR(author,0,4))`

ex : R-Romans - RAULT Antoine met à jour en **R RAU** dans la cote
ex : EA-Albums Jeunesse DENYS Joëlle met à jour en **EA DEN** dans la cote

**Comment by meteos77** [comment]
For documentaries / Pour les documentaires
to update the callnumber from dewey number and author / pour mettre à jour la cote (callnumber) depuis l'indice dewey (dewey_number) et auteur (author)
`UPDATE book SET callnumber = (deweynumber) || ' ' || SUBSTR(author,0,4))`

**Comment by meteos77** [comment]
Updated html (br) line break (CR) to keep different categories in export.
Mise à jour en html (br) du retour à la ligne (CR) pour garder les différentes catégories dans l'exportation.

Linux (\n)
`UPDATE book SET category = REPLACE(category, CHAR(10), '<br>')`

Windows (\r)
`UPDATE book SET category = REPLACE(category, CHAR(13), '<br>')`

**Comment by meteos77** [comment]
Use REGEXP to find documentaries (CALLNUMBER with numbers)

Utilisation des REGEXP pour avoir les documentaires (COTE avec des chiffres).
`SELECT callnumber,title FROM book WHERE callnumber REGEXP '(^[0-9])';`

Utilisation des REGEXP pour avoir les documentaires Jeunesse (COTE commençant par E puis des chiffres).
`SELECT callnumber,title FROM book WHERE callnumber REGEXP '(^E[0-9])';`

Utilisation des REGEXP pour avoir les documentaires (COTE avec des chiffres) donc le numéro d'acces contient 039000.
SELECT callnumber,title,accession_number,originality FROM book WHERE callnumber REGEXP '(^[0-9])' AND accession_number LIKE('039000%');

**Comment by meteos77** [comment]
Creation of a file for updating the callnumber for documentaries in the central library (039000%)
 / Creation d'un fichier de mise a jour de la Cote  pour les documentaires de la bibliothèque centrale (039000%)

`SELECT 'UPDATE book SET callnumber = "' || callnumber ||'"  WHERE accession_number = "' || accession_number || '" ;' FROM book WHERE callnumber REGEXP '(^[0-9])' AND accession_number LIKE('039000%') ORDER BY cast(accession_number as real) ASC ;`

The table created should be saved by selecting “File / Export table to CSV” (MAJCOTE_file.csv).
Le tableau crée est a sauvegarder par "Fichier / Exporter le tableau en CVS" (fichier_MAJCOTE.csv).

en bash
`sqlite3 "BDD_file.sqlite" < "MAJCOTE_file.csv"` / `sqlite3 "fichier_BDD.sqlite" < "fichier_MAJCOTE.csv"`
